pkg/compact/compact.go

```diff
@@ -214,6 +236,32 @@ func (c *Syncer) syncMetas(ctx context.Context) error {
 	return nil
 }
 
+func (c *Syncer) downloadMeta(ctx context.Context, id ulid.ULID) (*metadata.Meta, error) {
+	level.Debug(c.logger).Log("msg", "download meta", "block", id)
+
+	meta, err := block.DownloadMeta(ctx, c.logger, c.bkt, id)
+	if err != nil {
+		return nil, errors.Wrapf(err, "downloading meta.json for %s", id)
+	}
+
+	// ULIDs contain a millisecond timestamp. We do not consider blocks that have been created too recently to
+	// avoid races when a block is only partially uploaded. This relates to all blocks, excluding:
+	// - repair created blocks
+	// - compactor created blocks
+	// NOTE: It is not safe to miss "old" block (even that it is newly created) in sync step. Compactor needs to aware of ALL old blocks.
+	// TODO(bplotka): https://github.com/improbable-eng/thanos/issues/377
+	if ulid.Now()-id.Time() < uint64(c.syncDelay/time.Millisecond) &&
+		meta.Thanos.Source != metadata.BucketRepairSource &&
+		meta.Thanos.Source != metadata.CompactorSource &&
+		meta.Thanos.Source != metadata.CompactorRepairSource {
+
+		level.Debug(c.logger).Log("msg", "block is too fresh for now", "block", id)
```

Comment:
It's weird to return nil if error is nil. I would rather create sentinel error `errors.New("Block too young")` and check for this. What do you think?

Reply:
Yeah, I agree that it's a bit strange to return `nil, nil`.  

I think using a sentinel error as you suggest is better, I've updated to use that instead.